src/NHibernate/Type/CollectionType.cs

```diff
@@ -391,14 +392,36 @@ public virtual object InstantiateResult(object original)
 			// in which case we should never ever ever reset the dirty flag
 			// on the target because we simply do not know...
 			IPersistentCollection originalPc = original as IPersistentCollection;
-			IPersistentCollection resultPc = result as IPersistentCollection;
-			if(originalPc != null && resultPc!=null)
+			IPersistentCollection resultPc = target as IPersistentCollection;
+			if(resultPc!=null)
 			{
-				if (!originalPc.IsDirty)
+				if (elementsEqual || (originalPc != null && !originalPc.IsDirty))
```

Comment:
Well, maybe something like:

- Load an entity with a collection, change some elements in the collection.
- Instantiate a detached replica of the entity, with its collection filled with a copy of the modified one.
- Merge the detached entity into the session.

With your change, I think the entity would then have its collection flagged as non dirty, while it should be dirty.